fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitAggregateRule.java

```diff
@@ -152,49 +155,47 @@ private void checkDistinctAgg(List<CallOperator> distinctAggCallOperator) {
         }
     }
 
-    // True means good cases for two phase agg. It requires both group by keys and distinct cols are
-    // low cardinality which helps deduplication in the local stage.
-    private boolean hasAggregateEffect(OptExpression input, List<ColumnRefOperator> distinctColumns) {
-        Statistics statistics = input.getGroupExpression().getGroup().getStatistics();
-        Statistics inputStatistics = input.getGroupExpression().getInputs().get(0).getStatistics();
+    private boolean isTwoStageMoreEfficient(OptExpression input, List<ColumnRefOperator> distinctColumns) {
+        LogicalAggregationOperator aggOp = input.getOp().cast();
+        Statistics inputStatistics = input.getGroupExpression().inputAt(0).getStatistics();
         Collection<ColumnStatistic> inputsColumnStatistics = inputStatistics.getColumnStatistics().values();
-
-        if (inputsColumnStatistics.stream().anyMatch(ColumnStatistic::isUnknown)) {
+        if (inputsColumnStatistics.stream().anyMatch(ColumnStatistic::isUnknown) || !aggOp.hasLimit()) {
```

Comment:
why two stage is not efficient for agg without limit?

Reply:
The good case for two stage is the first phase using streaming mode and the second phase just need process little rows to calculate the aggregation result (it close to a streaming step). While the third stage need using a blocking step to complete the dereduplciate job, which leads it not efficient for two stage. But if without limit, the two stage also need a blocking step to calculate the final result then it cannot gain the benifit than three stage. On the contrary, the dereduplciate step in third stage can reduce the rows processed by the nest step can gain benifit.